File: lib/core/services/courses_service.dart
/// Courses Service
/// Handles course-related API calls

import '../api/api_client.dart';
import '../api/api_config.dart';
import '../api/api_response.dart';
import '../models/course_model.dart';

class CoursesService {
  final ApiClient _apiClient;

  CoursesService(this._apiClient);

  /// Get all courses with optional filters
  Future<ApiResponse<PaginatedResponse<Course>>> getCourses({
    int page = 1,
    int pageSize = 20,
    String? search,
    String? level,
    String? category,
    String? institutionId,
    bool? isOnline,
    bool? isActive,
  }) async {
    final queryParams = <String, dynamic>{
      'page': page,
      'page_size': pageSize,
      if (search != null) 'search': search,
      if (level != null) 'level': level,
      if (category != null) 'category': category,
      if (institutionId != null) 'institution_id': institutionId,
      if (isOnline != null) 'is_online': isOnline,
      if (isActive != null) 'is_active': isActive,
    };

    return await _apiClient.get(
      ApiConfig.courses,
      queryParameters: queryParams,
      fromJson: (data) => PaginatedResponse.fromJson(
        data,
        (json) => Course.fromJson(json),
      ),
    );
  }

  /// Get course by ID
  Future<ApiResponse<Course>> getCourseById(String courseId) async {
    return await _apiClient.get(
      '${ApiConfig.courses}/$courseId',
      fromJson: (data) => Course.fromJson(data),
    );
  }

  /// Create new course (Institution only)
  Future<ApiResponse<Course>> createCourse({
    required String title,
    required String description,
    required String level,
    required String category,
    required int duration,
    double? fee,
    String currency = 'USD',
    List<String> prerequisites = const [],
    required DateTime startDate,
    DateTime? endDate,
    int maxStudents = 100,
    bool isOnline = false,
  }) async {
    return await _apiClient.post(
      ApiConfig.courses,
      data: {
        'title': title,
        'description': description,
        'level': level,
        'category': category,
        'duration': duration,
        if (fee != null) 'fee': fee,
        'currency': currency,
        'prerequisites': prerequisites,
        'start_date': startDate.toIso8601String(),
        if (endDate != null) 'end_date': endDate.toIso8601String(),
        'max_students': maxStudents,
        'is_online': isOnline,
      },
      fromJson: (data) => Course.fromJson(data),
    );
  }

  /// Update course
  Future<ApiResponse<Course>> updateCourse({
    required String courseId,
    String? title,
    String? description,
    String? level,
    String? category,
    int? duration,
    double? fee,
    String? currency,
    List<String>? prerequisites,
    DateTime? startDate,
    DateTime? endDate,
    int? maxStudents,
    bool? isOnline,
    bool? isActive,
  }) async {
    return await _apiClient.patch(
      '${ApiConfig.courses}/$courseId',
      data: {
        if (title != null) 'title': title,
        if (description != null) 'description': description,
        if (level != null) 'level': level,
        if (category != null) 'category': category,
        if (duration != null) 'duration': duration,
        if (fee != null) 'fee': fee,
        if (currency != null) 'currency': currency,
        if (prerequisites != null) 'prerequisites': prerequisites,
        if (startDate != null) 'start_date': startDate.toIso8601String(),
        if (endDate != null) 'end_date': endDate.toIso8601String(),
        if (maxStudents != null) 'max_students': maxStudents,
        if (isOnline != null) 'is_online': isOnline,
        if (isActive != null) 'is_active': isActive,
      },
      fromJson: (data) => Course.fromJson(data),
    );
  }

  /// Delete course
  Future<ApiResponse<void>> deleteCourse(String courseId) async {
    return await _apiClient.delete('${ApiConfig.courses}/$courseId');
  }

  /// Get course statistics (Institution only)
  Future<ApiResponse<Map<String, dynamic>>> getCourseStats(String courseId) async {
    return await _apiClient.get(
      '${ApiConfig.courses}/$courseId/stats',
      fromJson: (data) => data as Map<String, dynamic>,
    );
  }

  /// Get enrolled students in course (Institution/Counselor)
  Future<ApiResponse<PaginatedResponse<Map<String, dynamic>>>> getEnrolledStudents({
    required String courseId,
    int page = 1,
    int pageSize = 20,
  }) async {
    return await _apiClient.get(
      '${ApiConfig.courses}/$courseId/students',
      queryParameters: {
        'page': page,
        'page_size': pageSize,
      },
      fromJson: (data) => PaginatedResponse.fromJson(
        data,
        (json) => json,
      ),
    );
  }

  /// Get recommended courses for student
  Future<ApiResponse<List<Course>>> getRecommendedCourses() async {
    return await _apiClient.get(
      '${ApiConfig.courses}/recommended',
      fromJson: (data) {
        final List<dynamic> list = data;
        return list.map((json) => Course.fromJson(json)).toList();
      },
    );
  }

  /// Search courses
  Future<ApiResponse<PaginatedResponse<Course>>> searchCourses({
    required String query,
    int page = 1,
    int pageSize = 20,
  }) async {
    return await _apiClient.get(
      '${ApiConfig.courses}/search',
      queryParameters: {
        'q': query,
        'page': page,
        'page_size': pageSize,
      },
      fromJson: (data) => PaginatedResponse.fromJson(
        data,
        (json) => Course.fromJson(json),
      ),
    );
  }
}

File: recommendation_service/app/services/courses_service.py
"""
Courses Service
Business logic for course management
"""
from typing import Optional, List, Dict, Any
from datetime import datetime
import logging
from uuid import uuid4

from app.database.config import get_supabase
from app.schemas.courses import (
    CourseCreateRequest,
    CourseUpdateRequest,
    CourseResponse,
    CourseListResponse,
    CourseStatistics,
    CourseStatus
)

logger = logging.getLogger(__name__)


class CoursesService:
    """Service for managing courses"""

    def __init__(self):
        self.db = get_supabase()

    async def create_course(
        self,
        institution_id: str,
        course_data: CourseCreateRequest
    ) -> CourseResponse:
        """Create a new course"""
        try:
            course = {
                "id": str(uuid4()),
                "institution_id": institution_id,
                "title": course_data.title,
                "description": course_data.description,
                "course_type": course_data.course_type.value,
                "level": course_data.level.value,
                "duration_hours": course_data.duration_hours,
                "price": course_data.price or 0.0,
                "currency": course_data.currency,
                "thumbnail_url": course_data.thumbnail_url,
                "preview_video_url": course_data.preview_video_url,
                "category": course_data.category,
                "tags": course_data.tags,
                "learning_outcomes": course_data.learning_outcomes,
                "prerequisites": course_data.prerequisites,
                "max_students": course_data.max_students,
                "syllabus": course_data.syllabus,
                "status": CourseStatus.DRAFT.value,
                "is_published": False,
                "enrolled_count": 0,
                "metadata": course_data.metadata or {},
                "created_at": datetime.utcnow().isoformat(),
                "updated_at": datetime.utcnow().isoformat(),
            }

            response = self.db.table('courses').insert(course).execute()

            if not response.data:
                raise Exception("Failed to create course")

            logger.info(f"Course created: {course['id']} by institution {institution_id}")

            return CourseResponse(**response.data[0])

        except Exception as e:
            logger.error(f"Create course error: {e}")
            raise Exception(f"Failed to create course: {str(e)}")

    async def get_course(self, course_id: str) -> CourseResponse:
        """Get course by ID"""
        try:
            response = self.db.table('courses').select('*').eq('id', course_id).single().execute()

            if not response.data:
                raise Exception("Course not found")

            return CourseResponse(**response.data)

        except Exception as e:
            logger.error(f"Get course error: {e}")
            raise Exception(f"Failed to fetch course: {str(e)}")

    async def update_course(
        self,
        course_id: str,
        institution_id: str,
        course_data: CourseUpdateRequest
    ) -> CourseResponse:
        """Update course"""
        try:
            # Verify ownership
            course = await self.get_course(course_id)
            if course.institution_id != institution_id:
                raise Exception("Not authorized to update this course")

            # Build update dict
            update_data = {"updated_at": datetime.utcnow().isoformat()}

            if course_data.title is not None:
                update_data["title"] = course_data.title
            if course_data.description is not None:
                update_data["description"] = course_data.description
            if course_data.course_type is not None:
                update_data["course_type"] = course_data.course_type.value
            if course_data.level is not None:
                update_data["level"] = course_data.level.value
            if course_data.duration_hours is not None:
                update_data["duration_hours"] = course_data.duration_hours
            if course_data.price is not None:
                update_data["price"] = course_data.price
            if course_data.currency is not None:
                update_data["currency"] = course_data.currency
            if course_data.thumbnail_url is not None:
                update_data["thumbnail_url"] = course_data.thumbnail_url
            if course_data.preview_video_url is not None:
                update_data["preview_video_url"] = course_data.preview_video_url
            if course_data.category is not None:
                update_data["category"] = course_data.category
            if course_data.tags is not None:
                update_data["tags"] = course_data.tags
            if course_data.learning_outcomes is not None:
                update_data["learning_outcomes"] = course_data.learning_outcomes
            if course_data.prerequisites is not None:
                update_data["prerequisites"] = course_data.prerequisites
            if course_data.max_students is not None:
                update_data["max_students"] = course_data.max_students
            if course_data.syllabus is not None:
                update_data["syllabus"] = course_data.syllabus
            if course_data.status is not None:
                update_data["status"] = course_data.status.value
                # If publishing, set published_at
                if course_data.status == CourseStatus.PUBLISHED and not course.is_published:
                    update_data["is_published"] = True
                    update_data["published_at"] = datetime.utcnow().isoformat()
            if course_data.metadata is not None:
                update_data["metadata"] = course_data.metadata

            response = self.db.table('courses').update(update_data).eq('id', course_id).execute()

            if not response.data:
                raise Exception("Failed to update course")

            logger.info(f"Course updated: {course_id}")

            return CourseResponse(**response.data[0])

        except Exception as e:
            logger.error(f"Update course error: {e}")
            raise Exception(f"Failed to update course: {str(e)}")

    async def delete_course(self, course_id: str, institution_id: str) -> Dict[str, str]:
        """Delete course (soft delete)"""
        try:
            # Verify ownership
            course = await self.get_course(course_id)
            if course.institution_id != institution_id:
                raise Exception("Not authorized to delete this course")

            # Soft delete by archiving
            self.db.table('courses').update({
                "status": CourseStatus.ARCHIVED.value,
                "updated_at": datetime.utcnow().isoformat()
            }).eq('id', course_id).execute()

            logger.info(f"Course deleted (archived): {course_id}")

            return {"message": "Course deleted successfully"}

        except Exception as e:
            logger.error(f"Delete course error: {e}")
            raise Exception(f"Failed to delete course: {str(e)}")

    async def list_courses(
        self,
        page: int = 1,
        page_size: int = 20,
        institution_id: Optional[str] = None,
        status: Optional[str] = None,
        category: Optional[str] = None,
        level: Optional[str] = None,
        search: Optional[str] = None,
        student_id: Optional[str] = None
    ) -> CourseListResponse:
        """List courses with filters and pagination (institution-restricted for students)"""
        try:
            # If student, filter by admitted institutions
            admitted_institution_ids = []
            if student_id:
                # Get institutions student is admitted to via accepted applications
                apps_response = self.db.table('applications')\
                    .select('programs!inner(institution_id)')\
                    .eq('student_id', student_id)\
                    .eq('status', 'accepted')\
                    .execute()

                if apps_response.data:
                    # Extract unique institution IDs
                    admitted_institution_ids = list(set([
                        app['programs']['institution_id']
                        for app in apps_response.data
                        if app.get('programs', {}).get('institution_id')
                    ]))

                # If student not admitted anywhere, return empty
                if not admitted_institution_ids:
                    return CourseListResponse(
                        courses=[],
                        total=0,
                        page=page,
                        page_size=page_size,
                        has_more=False
                    )

            query = self.db.table('courses').select('*', count='exact')

            # Apply filters
            if institution_id:
                query = query.eq('institution_id', institution_id)
            elif student_id and admitted_institution_ids:
                # Filter to only admitted institutions
                query = query.in_('institution_id', admitted_institution_ids)

            if status:
                query = query.eq('status', status)
            else:
                # By default, only show published courses
                query = query.eq('is_published', True)

            if category:
                query = query.eq('category', category)
            if level:
                query = query.eq('level', level)
            if search:
                query = query.ilike('title', f'%{search}%')

            # Pagination
            offset = (page - 1) * page_size
            query = query.order('created_at', desc=True).range(offset, offset + page_size - 1)

            response = query.execute()

            courses = [CourseResponse(**course) for course in response.data] if response.data else []
            total = response.count or 0

            return CourseListResponse(
                courses=courses,
                total=total,
                page=page,
                page_size=page_size,
                has_more=(offset + page_size) < total
            )

        except Exception as e:
            logger.error(f"List courses error: {e}")
            raise Exception(f"Failed to list courses: {str(e)}")

    async def publish_course(self, course_id: str, institution_id: str) -> CourseResponse:
        """Publish a course"""
        try:
            course = await self.get_course(course_id)
            if course.institution_id != institution_id:
                raise Exception("Not authorized to publish this course")

            update_data = {
                "status": CourseStatus.PUBLISHED.value,
                "is_published": True,
                "published_at": datetime.utcnow().isoformat(),
                "updated_at": datetime.utcnow().isoformat()
            }

            response = self.db.table('courses').update(update_data).eq('id', course_id).execute()

            if not response.data:
                raise Exception("Failed to publish course")

            logger.info(f"Course published: {course_id}")

            return CourseResponse(**response.data[0])

        except Exception as e:
            logger.error(f"Publish course error: {e}")
            raise Exception(f"Failed to publish course: {str(e)}")

    async def unpublish_course(self, course_id: str, institution_id: str) -> CourseResponse:
        """Unpublish a course"""
        try:
            course = await self.get_course(course_id)
            if course.institution_id != institution_id:
                raise Exception("Not authorized to unpublish this course")

            update_data = {
                "status": CourseStatus.DRAFT.value,
                "is_published": False,
                "updated_at": datetime.utcnow().isoformat()
            }

            response = self.db.table('courses').update(update_data).eq('id', course_id).execute()

            if not response.data:
                raise Exception("Failed to unpublish course")

            logger.info(f"Course unpublished: {course_id}")

            return CourseResponse(**response.data[0])

        except Exception as e:
            logger.error(f"Unpublish course error: {e}")
            raise Exception(f"Failed to unpublish course: {str(e)}")

    async def get_course_statistics(self, institution_id: str) -> CourseStatistics:
        """Get course statistics for an institution"""
        try:
            # Get all courses for institution
            response = self.db.table('courses').select('*').eq('institution_id', institution_id).execute()

            courses = response.data if response.data else []

            stats = {
                "total_courses": len(courses),
                "published_courses": sum(1 for c in courses if c.get('status') == CourseStatus.PUBLISHED.value),
                "draft_courses": sum(1 for c in courses if c.get('status') == CourseStatus.DRAFT.value),
                "archived_courses": sum(1 for c in courses if c.get('status') == CourseStatus.ARCHIVED.value),
                "total_enrollments": sum(c.get('enrolled_count', 0) for c in courses),
                "average_rating": 0.0,  # TODO: Calculate from reviews
                "total_revenue": sum(c.get('price', 0) * c.get('enrolled_count', 0) for c in courses)
            }

            return CourseStatistics(**stats)

        except Exception as e:
            logger.error(f"Get statistics error: {e}")
            raise Exception(f"Failed to get statistics: {str(e)}")
